onnxruntime/core/providers/vitisai/imp/ep_context_utils.cc

```diff
@@ -0,0 +1,345 @@
+    auto& meta_def_inputs = p_meta_def->inputs();
+    for (int i = 0; i < isg_meta_def_inputs_size; i++) {
+      meta_def_inputs.push_back(p_func_proto->input(i));
+    }
+    auto& meta_def_outputs = p_meta_def->outputs();
+    for (int i = 0, l = p_func_proto->output_size(); i < l; i++) {
+      meta_def_outputs.push_back(p_func_proto->output(i));
+    }
+    auto& meta_def_initializers = p_meta_def->constant_initializers();
+    for (int i = isg_meta_def_inputs_size, l = p_func_proto->input_size(); i < l; i++) {
+      meta_def_initializers.push_back(p_func_proto->input(i));
+    }
+    auto& meta_def_attrs = p_meta_def->attributes();
+    for (int i = 0, l = p_func_proto->attribute_size(); i < l; i++) {
+      meta_def_attrs.emplace(p_func_proto->attribute(i), p_func_proto->attribute_proto(i));
+    }
+    p_meta_def->doc_string() = p_func_proto->doc_string();
+    // TODO: `IndexedSubGraph::type_and_shape_inference_function`.
+    p_isg->SetMetaDef(std::move(p_meta_def));
+  }
+  auto& isg_nodes = p_isg->Nodes();
+  for (int i = 0, l = p_func_proto->node_size(); i < l; i++) {
+    const auto& node_proto = p_func_proto->node(i);
+    isg_nodes.push_back(node_proto.attribute(const_cast<ONNX_NAMESPACE::NodeProto&>(node_proto).attribute_size() - 1).i());
+  }
+  auto schema_source = static_cast<IndexedSubGraph_SourceOfSchema>(
+      std::stoi(*(const_cast<ONNX_NAMESPACE::StringStringEntryProto&>(p_func_proto->metadata_props(func_metadata_props_size - 1)).mutable_value())));
+  p_isg->SetSchemaSource(schema_source);
+  return p_isg;
+}
+
+std::string SerializeCapabilities(
+    const std::vector<std::unique_ptr<ComputeCapability>>& capability_ptrs,
+    const Graph& graph) {
+  std::stringstream ss;
+  for (const auto& p : capability_ptrs) {
+    auto& p_subgraph = p->SubGraph();
+    auto p_func_proto = ConvertIndexedSubGraphToFunctionProto(*p_subgraph, graph);
+    std::string func_proto_buf;
+    p_func_proto->SerializeToString(func_proto_buf);
+    size_t buf_len = func_proto_buf.length();
+    ss.write(reinterpret_cast<const char*>(&buf_len), sizeof(buf_len));
+    ss.write(func_proto_buf.data(), buf_len);
+  }
+  if (!ss.good()) {
+    ORT_THROW("Serialization stream bad");
+  }
+  return ss.str();
+}
+
+void DeserializeCapabilities(const std::string& ser_capabilities,
+                             std::vector<std::unique_ptr<ComputeCapability>>& capability_ptrs) {
+  std::istringstream ss(ser_capabilities);
+  while (!ss.eof()) {
+    size_t buf_len;
+    ss.read(reinterpret_cast<char*>(&buf_len), sizeof(buf_len));
+    std::string buf(buf_len, '\0');
+    ss.read(&buf[0], buf_len);
+    auto p_func_proto = ONNX_NAMESPACE::FunctionProto::Create();
+    p_func_proto->ParseFromString(buf);
+    auto p_subgraph = ConvertFunctionProtoToIndexedSubGraph(p_func_proto);
+    capability_ptrs.push_back(ComputeCapability::Create(std::move(p_subgraph)));
+  }
+}
+
+std::unique_ptr<Model> CreateEPContexModel(
```

Comment:
Many places directly process the proto_buf, the existing API like graph.AddNode, graph.GetOrCreateNodeArg, node.AddAttribute, etc should be enough. Any special reason to add new provider interface like FunctionProto?
Please refer to code example here:
https://github.com/microsoft/onnxruntime/blob/3ecb01233789d4057a902d3e559ac6a9b410cbed/onnxruntime/core/providers/qnn/builder/onnx_ctx_model_helper.cc#L189-L251

Reply:
Inspired by the comments of `onnxruntime::IndexedSubGraph` at https://github.com/microsoft/onnxruntime/blob/3ecb01233789d4057a902d3e559ac6a9b410cbed/include/onnxruntime/core/graph/indexed_sub_graph.h#L28-L68, we are serializing an `IndexedSubGraph` (which is essentially the concrete content of `ComputeCapbility`) to a `FunctionProto`.
However, `FunctionProto` is missing from the file "provider_wrappedtypes.h" (as well as the corresponding backing files). So, we are doing the current way.

Reply:
In terms of the procedure/function of creating an EP context model, actually it's very like the one you referred to of QNN EP.